### PR TITLE
Create prefix matches for wildcard SPIFFE URI SANs in TLS validation context

### PR DESCRIPTION
**Pull request by kailun-qin**
This helps with cases e.g. SPIFFE where users would like to trust the entire
trust domain by setting wildcards SANs like "spiffe://cluster.domain/*".

Signed-off-by: Kailun Qin <kailun.qin@intel.com>

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/security/model/authentication.go
Comment (ramaraochavali) [inline pilot/pkg/security/model/authentication.go:269]: https://github.com/istio/api/blob/master/networking/v1alpha3/destination_rule.proto#L948 - based on this doc, I think we assume it is exact match. I get the intent of the PR, can  we clarify/update docs before we merge this?

Comment (kailun-qin) [reply pilot/pkg/security/model/authentication.go:269]: Thanks for the comments! Proposing doc updates here: https://github.com/istio/api/pull/1966.

Comment (kailun-qin) [reply pilot/pkg/security/model/authentication.go:269]: Good catch, thanks! Right, should applied to gateway, updated. Also attaching a release note.

File: releasenotes/notes/san-wildcard.yaml
Comment (shamsher31) [inline releasenotes/notes/san-wildcard.yaml:7]:
```suggestion
  **Added** support for subject alternative names that end with wildcards in TLS validation context. This allows users to specify `subjectAltNames` end in wildcards (e.g., "spiffe://cluster.domain/*") for gateway `ServerTLSSettings`.
```

Comment (kailun-qin) [reply releasenotes/notes/san-wildcard.yaml]: Updated, thanks!

File: pilot/pkg/networking/util/util.go
Comment (howardjohn) [inline pilot/pkg/networking/util/util.go:631]: My memory here is a bit fuzzy and I didn't dig up the RFC, but my memory of SAN matching is:
* Prefix match is not allowed
* Suffix match only goes one level deep. So `*.foo.com` should not match `a.b.foo.com`. There was some envoy CVE about this

**Review by howardjohn — CHANGES_REQUESTED** [review]
I am a bit concerned that this will allow users to do strange things. While the spiffe case may make sense in isolation, the general API of allowing suffix matches seems suspicious and inconsistent (no prefix match and DR is missing?).

File: releasenotes/notes/san-wildcard.yaml
Comment (kailun-qin) [reply releasenotes/notes/san-wildcard.yaml:6]: I think this goes to DR as well (just the current impl. impacts gw only), as long as the wildcard use case is reasonable/acceptable.

File: pilot/pkg/networking/util/util.go
Comment (kailun-qin) [reply pilot/pkg/networking/util/util.go:631]: Thanks for the summary! They're mostly correct IMO (though not clear/consistent amongst different impl.).

However, usually this only applies to `subjectAltName`s of type DNS name (or similar). For an ID of type `application service`, e.g., SPIFFE URI, there seems to be no such standard.

https://tools.ietf.org/html/rfc6125#section-6 might be a good reference.

Comment (howardjohn) [reply pilot/pkg/networking/util/util.go:631]: Yeah that makes sense, but this API can be used for either so its a bit tricky. We need to be careful about being so permissive that we allow people to do things that are not safe. 

@myidpt @costinm may have some opinions here

**Comment by kailun-qin** [comment]
> I am a bit concerned that this will allow users to do strange things. While the spiffe case may make sense in isolation, the general API of allowing suffix matches seems suspicious and inconsistent (no prefix match and DR is missing?).

Thanks for the comments! The concerns are reasonable IMHO. Regarding suspicious suffix/prefix matching, according to the RFC cited, this usually applies to DNS(-like) type of ID - and we don't intend to support this due to security/standard concerns. Regarding DR support, I think this (SPIFFE URI validation) can apply to DR as well.

Can we support this in a more narrow scope - e.g., SPIFFE URI only, with defined matching?

cc @ramaraochavali, I noticed https://github.com/envoyproxy/envoy/issues/3883 and https://github.com/envoyproxy/envoy/issues/7813 might be related.

**Comment by ramaraochavali** [comment]
Another way is to add another subject_alt_names https://github.com/istio/api/blob/master/networking/v1alpha3/virtual_service.proto#L1484 - call it match_subject_alt_names and there we can support prefix as Envoy supports?

**Comment by kailun-qin** [comment]
Limit the scope of wildcard support to SPIFFE URI only and apply to DR as well. PTAL, thanks! @howardjohn

File: pilot/pkg/security/model/authentication.go
Comment (howardjohn) [inline pilot/pkg/security/model/authentication.go:301]: do we want exact match AND prefix match? I would have thought its one or the other

File: pilot/pkg/security/model/authentication.go
Comment (howardjohn) [inline pilot/pkg/security/model/authentication.go:304]: the API docs said `spiffe://foo.cl*` was not allowed but I don't see anywhere that validates that?

**Review by howardjohn — COMMENTED** [review]
@costinm @incfly ptal

File: pilot/pkg/security/model/authentication.go
Comment (kailun-qin) [reply pilot/pkg/security/model/authentication.go:301]: Agree, I think one or the other would suffice.

File: pilot/pkg/security/model/authentication.go
Comment (kailun-qin) [reply pilot/pkg/security/model/authentication.go:304]: `spiffe://foo.cl*` would anyway create an exact match but never a prefix match, this is what Istio does now (no check on SPIFFE URI SAN, and just applies) and the one which is called `allowed`. It is guaranteed by only create one prefix match when with `/*` suffix.

File: pilot/pkg/security/model/authentication.go
Comment (kailun-qin) [reply pilot/pkg/security/model/authentication.go]: Updated, thanks!

**Review by incfly — CHANGES_REQUESTED** [review]
Put up some thoughts, but I may miss some previous context.

---

IMO this requires a formal security WG approval for such behavior change. If so, can you point to some reference? design doc or security WG meeting notes.

- Need more specific use cases, and reason why current API configuration is not feasible.
- Once added and add more semantics to API, it'll be very hard to remove. Surface area for CVE also becomes larger. e.g is `*` an invalid character in spiffe spec? Does this wildcard a known notion in SPIFFE or we are inventing some syntax for our own? 

`spiffe://foo.com/*` would effectively trust all the workloads under the same trust domain. Therefore any compromised workloads in in that trust domain can become MITM to fool the client. Without design doc, it's hard to tell whether this is intended for temporary use case (for migration etc), or end up encouraging users to stay with this due to conveniences.

**Comment by kailun-qin** [comment]
> Put up some thoughts, but I may miss some previous context.
> 
> IMO this requires a formal security WG approval for such behavior change. If so, can you point to some reference? design doc or security WG meeting notes.
> 
> * Need more specific use cases, and reason why current API configuration is not feasible.
> * Once added and add more semantics to API, it'll be very hard to remove. Surface area for CVE also becomes larger. e.g is `*` an invalid character in spiffe spec? Does this wildcard a known notion in SPIFFE or we are inventing some syntax for our own?
> 
> `spiffe://foo.com/*` would effectively trust all the workloads under the same trust domain. Therefore any compromised workloads in in that trust domain can become MITM to fool the client. Without design doc, it's hard to tell whether this is intended for temporary use case (for migration etc), or end up encouraging users to stay with this due to conveniences.

Thanks for the comments!
Make sense to me. Let me see if I can add w/ more use cases and ask for another review at the WG meeting.

**Review by costinm — COMMENTED** [review]
looks good to me.

**Pull request closed without merging**
